csrc/runtime/executor_kernel_arg.h

```diff
@@ -51,30 +51,68 @@ class KernelArgumentHolder {
 
   NVF_API void push(const std::vector<at::Tensor>& tensors);
 
+  void erase(const PolymorphicValue& arg_to_delete);
+
   c10::ArrayRef<c10::IValue> toArrayRef() const;
 
   void erase(const PolymorphicValue* arg_to_delete);
 
   void push(PolymorphicValue val) {
-    arguments_.push_back(std::make_shared<PolymorphicValue>(std::move(val)));
+    arguments_.push_back(PolymorphicValue(std::move(val)));
+  }
+
+  PolymorphicValue& back() {
+    return arguments_.back();
   }
 
-  PolymorphicValue* back() {
-    return arguments_.back().get();
+  const PolymorphicValue& back() const {
+    return arguments_.back();
   }
 
-  PolymorphicValue* operator[](size_t ind) const {
-    return arguments_.at(ind).get();
-  };
+  PolymorphicValue& operator[](size_t ind) {
+    return arguments_.at(ind);
+  }
 
+  const PolymorphicValue& operator[](size_t ind) const {
+    return arguments_.at(ind);
+  }
+
+  // Returns iterator pointing to the beginning of vector container
   auto begin() const {
     return arguments_.begin();
   }
 
+  // Returns iterator pointing to the end of vector container
   auto end() const {
     return arguments_.end();
   }
 
+  // Returns iterator pointing to the beginning of vector container
```

Comment:
Just building out some vector like functionality as this class is basically a vector of polymorphic values